### PR TITLE
Add abstractType for the resource node match

### PR DESCRIPTION
**Pull request by jiaolu**
Usually matchType equals resource node type, in some special case,
matchType would be converted from other sources.
this PR is related with https://github.com/GPUOpen-Drivers/llpc/pull/1915

### CODE REVIEW

**Comment by qiaojbao** [comment]
retest this please

**Comment by qiaojbao** [comment]
16:04:00  [6784/73567] Pass: 618 Fail: 2484 Skip: 2570 ExpectedFail: 0 UnexpectedPass: 0 Crash: 24 Timeout: 0 Missing: 874 Flake: 214 Duration: 23:48 Remaining: 3:54:19
16:04:01  [0/73567] Pass: 0 Fail: 0 Skip: 0 ExpectedFail: 0 UnexpectedPass: 0 Crash: 0 Timeout: 0 Missing: 0 Flake: 0 Duration: 23:46
16:04:01  [0/69153] Pass: 0 Fail: 0 Skip: 0 ExpectedFail: 0 UnexpectedPass: 0 Crash: 0 Timeout: 0 Missing: 0 Flake: 0 Duration: 23:52

**Comment by nhaehnle** [comment]
Can you please give a concrete example where those types would be different and why (e.g. source SPIR-V)?

**Comment by jiaolu** [comment]
> Can you please give a concrete example where those types would be different and why (e.g. source SPIR-V)?

this is for hlsl binding semantics,  we can talk about it llpc-chat

**Comment by qiaojbao** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 603756201512843b0cd026e7a12953eeac318e11
<details><summary>CTS tests (Failed: 0/216287)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35463/73567 (48.2%)</li>
<li>Failed: 0/73567 (0.0%)</li>
<li>Not Supported: 38104/73567 (51.8%)</li><li>Warnings: 0/73567 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33156/69153 (47.9%)</li>
<li>Failed: 0/69153 (0.0%)</li>
<li>Not Supported: 35997/69153 (52.1%)</li><li>Warnings: 0/69153 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37728/73567 (51.3%)</li>
<li>Failed: 0/73567 (0.0%)</li>
<li>Not Supported: 35839/73567 (48.7%)</li><li>Warnings: 0/73567 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/interface/lgc/Pipeline.h [hunk 1]
```diff
@@ -241,9 +241,10 @@ struct ShaderOptions {
 struct ResourceNode {
   ResourceNode() {}
 
-  ResourceNodeType type;   // Type of this node
-  unsigned sizeInDwords;   // Size in dwords
-  unsigned offsetInDwords; // Offset in dwords
+  ResourceNodeType type;      // Type of this node
```
Comment (s-perron) [inline lgc/interface/lgc/Pipeline.h:244]: The names `type` and `matchType` are a little confusing.  Could we rename `type` to something like `underlyingType` and then rename `matchType` to `resourceType`.  The problem  with a name like `matchType` is that the name says what you will do with it, but not what it is.  A variable name is best when it describes the data is contains, and the code can use it in appropriate ways.

Comment to describe the difference between the two will be useful as well.

File: llpc/context/llpcPipelineContext.cpp [hunk 2]
```diff
@@ -550,6 +550,7 @@ void PipelineContext::setUserDataNodesTable(Pipeline *pipeline, ArrayRef<Resourc
 
     destNode.sizeInDwords = node.sizeInDwords;
     destNode.offsetInDwords = node.offsetInDwords;
+    destNode.matchType = ResourceNodeType::Unknown;
```
Comment (s-perron) [inline llpc/context/llpcPipelineContext.cpp:553]: I think it would be best to always have an appropriate value in `matchType` (or whatever it gets renamed to).  Why not have the `matchType` contain `ResourceMappingNodeType::DescriptorTableVaPtr` when it is a descriptor table?

**Review by s-perron — COMMENTED** [review]
It is hard the review if this is the right way to do things when I cannot see how it will be used.  Can you also make changes to the code that finds the resource nodes?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0b9045cb48e0b239fc72104352cffc0e6212a9cd
<details><summary>CTS tests (Failed: 1/216059)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35297/73410 (48.1%)</li>
<li><details><summary>Failed: 1/73410 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.buffer_marker.compute.external_host_mem.bottom_of_pipe.memory_dep.dispatch
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 38112/73410 (51.9%)</li><li>Warnings: 0/73410 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33017/69239 (47.7%)</li>
<li>Failed: 0/69239 (0.0%)</li>
<li>Not Supported: 36222/69239 (52.3%)</li><li>Warnings: 0/69239 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73410 (51.2%)</li>
<li>Failed: 0/73410 (0.0%)</li>
<li>Not Supported: 35857/73410 (48.8%)</li><li>Warnings: 0/73410 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/interface/lgc/Pipeline.h [hunk 3]
```diff
@@ -138,10 +138,10 @@ struct Options {
...
   unsigned useResourceBindingRange;    // A resource node binding is the start of a range whose size is
                                        //  sizeInDwords/stride.
-  unsigned reserved1f;                // Reserved for future functionality
-  unsigned enableInterpModePatch; // Enable to do per-sample interpolation for nonperspective and smooth input
-  unsigned pageMigrationEnabled;  // Enable page migration
-  ResourceLayoutScheme resourceLayoutScheme; // Resource layout scheme
+  unsigned reserved1f;                 // Reserved for future functionality
+  unsigned enableInterpModePatch;      // Enable to do per-sample interpolation for nonperspective and smooth input
+  unsigned pageMigrationEnabled;       // Enable page migration
+  ResourceLayoutScheme resourceLayoutScheme;     // Resource layout scheme
```
Comment (nhaehnle) [inline lgc/interface/lgc/Pipeline.h:144]: Unrelated whitespace change?

**Review by nhaehnle — APPROVED** [review]
LGTM

**Review by s-perron — APPROVED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9c06e40d9df8f000cbf9939c69e36a7cad96c44d
<details><summary>CTS tests (Failed: 0/216059)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35298/73410 (48.1%)</li>
<li>Failed: 0/73410 (0.0%)</li>
<li>Not Supported: 38112/73410 (51.9%)</li><li>Warnings: 0/73410 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33017/69239 (47.7%)</li>
<li>Failed: 0/69239 (0.0%)</li>
<li>Not Supported: 36222/69239 (52.3%)</li><li>Warnings: 0/69239 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73410 (51.2%)</li>
<li>Failed: 0/73410 (0.0%)</li>
<li>Not Supported: 35857/73410 (48.8%)</li><li>Warnings: 0/73410 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9c06e40d9df8f000cbf9939c69e36a7cad96c44d
<details><summary>CTS tests (Failed: 0/216059)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35298/73410 (48.1%)</li>
<li>Failed: 0/73410 (0.0%)</li>
<li>Not Supported: 38112/73410 (51.9%)</li><li>Warnings: 0/73410 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33017/69239 (47.7%)</li>
<li>Failed: 0/69239 (0.0%)</li>
<li>Not Supported: 36222/69239 (52.3%)</li><li>Warnings: 0/69239 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37553/73410 (51.2%)</li>
<li>Failed: 0/73410 (0.0%)</li>
<li>Not Supported: 35857/73410 (48.8%)</li><li>Warnings: 0/73410 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]
LGTM, needs a squash merge.
